website/blog/2022-10-18-how-to-design-metrics.md

```diff
@@ -0,0 +1,367 @@
+| 2022-10-02 | 5 | nps_tool_2 | developer | team | 9 | promoter | 100 |
+| 2022-10-03 | 6 | nps_tool_1 | developer | enterprise | 7 | passive | 0 |
+
+Now that each survey has an associated `nps_value` we can forgo the ratio calculations used in the Metric Logic section and create our NPS Score metric as a single average metric.
+
+```yaml
+metrics:
+  - name: nps_score
+    label: NPS Score
+    model: ref('customer_nps')
+    calculation_method: average
+    expression: nps_value
+    timestamp: created_at
+    time_grains: [day, month, quarter, year]
+    dimensions:
+      - feedback_source
+      - account_plan
+      - user_type
+```
+    
+<details>
+  <summary>Why does this work?</summary>
+
+This is a slightly different way of calculating NPS from the usually provided formula but it ends up with the same result. Here is why:
+
+- `promoter_pct` was defined as `total_promoter_respondents` / `total_respondents`
+    - In our example dataset, this nets out to 3 / 6 = 50%.
+    - If we instead assign a value of 100 and take the average, it becomes 300 / 6 = 50.
+- `detractor_pct` was defined as `total_detractor_respondents` / `total_respondents`
+    - In our example dataset, this nets out to 1 / 6 = 16.67%.
+    - If we instead assign a value of 100 and take the average, it becomes -100 / 6 = -16.67.
+- Therefore, our `nps_score` follows suit:
+    - In our example dataset, 50% - 16.67% = 33.33%
+    - If we instead assign a value of 100 and take the average, it becomes 200 / 6 = 33.33
+
+The underlying principle of why this works is based on the fact that averages divide the sum of the values in the set by their number. In more dbt friendly terms, what it really means is that average is creating the following equation: `sum(value)/count(*)`. In the first example implementation, we were doing roughly the same thing with multiple metric definitions - the only difference was our numerator was a count that assigned each row a value of 1. So if we duplicate that logic and give each row a value of 1 then we can create far fewer metrics.
+
+But that only gets us to the `promoter_pct` and `detractor_pct` metrics. In order to combine these both into a single metric definition, we needed to change the value that we assign. Given that the total range of values that the metric could output is -100 (all detractors) to 100 (all promoters) we can assign each of those categories that peak value, along with 0 for passives. This means that when the numerator is aggregated, it nets out promoters against detractors just like the documented equation does `promoter score - detractor score` . 
+
+</details>
+
+**Is this what I should do?**
+
+It depends! There will be times when it might be better to have logic stored in the modeling layer and there will be times when it might be better to have logic stored in the semantic layer. The needs of your business stakeholders should drive your decision on where to keep this logic.
```

Comment:
1. Should we link this off of the "it depends" https://twitter.com/SBinLondon/status/1413113782214266885
2. Can we give some flavor to how folks might choose between the "it depends"

Reply:
1. Yea!
2. I honestly don't know if we/me yet have opinions on when to do each option. The best short hand I've gotten has been "Put it all in the semantic layer if you want to analyze/use the individual components as well"

Reply:
Added a comment here